components/sim800l.rst

```diff
@@ -0,0 +1,156 @@
+connected via a :doc:`UART bus </components/uart>` supporting both receiving and transmitting line. 
+The uart bus must be configured at the same speed of the module which is by default 9600bps. 
+The required connection wires are `+VCC`, `GND`, `RX` and `TX`.
+
+.. warning::
+
+    If you are using the :doc:`logger` make sure you are not using the same pins for ``TX`` and ``RX`` or
+    otherwise disable the uart logging with the ``baud_rate: 0`` option.
+
+.. note::
+
+    This module requires a power supply between 3.8V and 4.2V that can handle current spikes up
+    to 2 amps, it will not work by powering from the same 3.3v power source of the ESP. However you can
+    connect `TX` and `RX` lines directly without any level shifter.
+
+.. figure:: images/sim800l-full.jpg
+    :align: center
+    :width: 60.0%
+
+.. code-block:: yaml
+
+    # Example configuration entry
+    uart:
+      baud_rate: 9600
+      tx_pin: TX
+      rx_pin: RX
+
+    sim800l:
+      id: sim800l1
+      on_sms_received:
+        - logger.log:
+            format: "Received '%s' from %s"
+            args: [ 'message.c_str()', 'sender.c_str()' ]
+
+Configuration variables:
+------------------------
+
+- **uart_id** (*Optional*, :ref:`config-id`): Manually specify the ID of the UART hub.
+- **id** (*Optional*, :ref:`config-id`): Manually specify the ID used for code generation.
+- **on_sms_received** (*Optional*, :ref:`Automation <automation>`): An action to be
+  performed when a sms is received. See :ref:`sim800l-on_sms_received`.
+
+.. _sim800l-on_sms_received:
+
+``on_sms_received`` Trigger
+---------------------------
+
+With this configuration option you can write complex automations whenever a sms message
+is received. To use the message content, use a :ref:`lambda <config-lambda>`
+template, the message content and the sender phone number are available inside that lambda
+under the variables named ``message`` and ``sender`` respectively.
+
+.. code-block:: yaml
+
+  on_sms_received:
+    - lambda: |-
+        id(sms_sender).publish_state(sender);
+        id(sms_message).publish_state(message);
+
+
+.. _sim800l-send_sms_action:
+
+``sim800l.send_sms`` Action
+---------------------------
+
+Send a SMS message to a phone recipient using this action in automations.
+
+.. code-block:: yaml
+
+    on_...:
+      then:
+        - sim800l.send_sms:
+            id: sim800l1
+            recipient: '+15551234567'
+            message: Hello there
+
+        # Templated:
+        - sim800l.send_sms:
+            id: sim800l1
+            recipient: !lambda |-
+              if (id(reed_switch).state) return "+15551234567";
+              else return "15551234568";
+            message: !lambda |-
+              return id(reed_switch).state ? "Door is now OPEN" : "Hey door just CLOSED";
+
+Configuration options:
+
+- **id** (*Required*, string): The ID of the SIM800L
```

Comment:
See https://github.com/esphome/esphome/pull/522#discussion_r288912605 - this should not be a required option. Make it optional and move it to the bottom of the configuration options.

```suggestion
- **id** (*Optional*, :ref:`config-id`): The ID of the SIM800L module to send the SMS with,
  defaults to the first one in the config.
```